### ISSUE TITLE
Convention for accessing TLS certificates from protocols

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7503]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#313|
|Type|enhancement|
|Created|2003-10-04 12:20:04Z|

<details><summary>Searchable metadata</summary>

```
trac-id__313 313
type__enhancement enhancement
reporter__exarkun exarkun
priority__low low
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__wontfix wontfix
component__core core
keywords__ 
time__1065270004000000 1065270004000000
changetime__1468259369629439 1468259369629439
version__ 
owner__ 
cc__exarkun cc__itamarst cc__moshez
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

```
#!html
<pre>
strports should be able to specify certificate information for
negotiate-after-connect TLS as well as plain old SSL.  This can be acheived by
forcing protocols to access their certificate in a particular way.

&lt;itamar> "any protocol sufficiently sophisticated to require STARTTLS will also
move enough functionality into a portal that you can just attach same portal to
multiple factories and it's like listening with same factory on multiple ports"
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="moshez's avatar" src="https://avatars.githubusercontent.com/u/422703?s=50" width="50" height="50">](https://github.com/moshez)|@moshez commented|
|-|-|

```
#!html
<pre>
On Sat, 04 Oct 2003, Jp Calderone wrote:

> strports should be able to specify certificate information for
> negotiate-after-connect TLS as well as plain old SSL. This can be acheived
> by forcing protocols to access their certificate in a particular way.

Assuming it's self.factory.contextFactory, I have included a preliminary
patch to strports. I'll let someone else deal with the other repercussions,
but I will list the ones I see here

* Make sure everything included in Twisted supports this
* Clearly document it in IServerFactory or something's docstring
* Document it in servers.html

Enjoy!

**************************************
Index: twisted/application/strports.py
===================================================================
RCS file: /cvs/Twisted/twisted/application/strports.py,v
retrieving revision 1.1
diff -u -r1.1 strports.py
--- twisted/application/strports.py     26 Sep 2003 22:44:48 -0000      1.1
+++ twisted/application/strports.py     4 Oct 2003 05:17:25 -0000
@@ -46,15 +46,20 @@
 """
 from __future__ import generators

-def _parseTCP(factory, port, interface="", backlog=5):
+def _parseTCP(factory, port, interface="", backlog=5,
+              privateKey=None, certKey=None):
+    if privateKey:
+        cf = _makeContextFactory(certKey, privateKey)
+        factory.contextFactory = cf
     return (int(port), factory), {'interface': interface,
                                   'backlog': int(backlog)}

-def _parseUNIX(factory, address, mode='666', backlog=5):
+def _parseUNIX(factory, address, mode='666', backlog=5,
+               privateKey=None, certKey=None):
+    cf = _makeContextFactory(certKey, privateKey)
     return (address, factory), {'mode': int(mode, 8), 'backlog': int(backlog)}

-def _parseSSL(factory, port, privateKey="server.pem", certKey=None,
-              sslmethod=None, interface='', backlog=5):
+def _makeContextFactory(certKey, privateKey):
     from twisted.internet import ssl
     if certKey is None:
         certKey = privateKey
@@ -62,6 +67,11 @@
     if sslmethod is not None:
         kw['sslmethod'] = getattr(ssl.SSL, sslmethod)
     cf = ssl.DefaultOpenSSLContextFactory(privateKey, certKey, **kw)
+    return cf
+
+def _parseSSL(factory, port, privateKey="server.pem", certKey=None,
+              sslmethod=None, interface='', backlog=5):
+    cf = _makeContextFactory(certKey, privateKey)
     return ((int(port), factory, cf),
             {'interface': interface, 'backlog': int(backlog)})

@@ -134,11 +144,14 @@

     For TCP, the arguments are the port (port number) and, optionally the
     interface (interface on which to listen) and backlog (how many clients to
-    keep in the backlog).
+    keep in the backlog). Optionally, privateKey and certKey can be specified,
+    in which case the server will use TLS if it supports it.

     For UNIX domain sockets, the arguments are address (the file name of the
     socket) and optionally the mode (the mode bits of the file, as an octal
     number) and the backlog (how many clients to keep in the backlog).
+    Optionally, privateKey and certKey can be specified, in which case the
+    server will use TLS if it supports it.

     For SSL sockets, the arguments are the port (port number) and, optionally,
     the privateKey (file in which the private key is in), certKey (file in
**************************************
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="moshez's avatar" src="https://avatars.githubusercontent.com/u/422703?s=50" width="50" height="50">](https://github.com/moshez)|@moshez commented|
|-|-|

```
#!html
<pre>
Re: multiple factories.
I think the behaviour of strports should be in those cases to first copy the 
factory and then to mutate it. But the patch is largely correct, so
strportifying shouldn't be an issue
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set status to `closed`|
|-|-|

Nevermind.  I am no longer enamoured of strports.  I withdraw this feature request.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="GitHub's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|GitHub commented|
|-|-|

In changeset bd34eb3cb09791f4c7eecb831a468743b2fe52b1
```
#!CommitTicketReference repository="" revision="bd34eb3cb09791f4c7eecb831a468743b2fe52b1"
Merge pull request #313 from twisted/8584-rodrigc-tcp

Author: rodrigc
Reviewer: adiroiban
Fixes: #8584

Increase test coverage of test_tcp_internals.py on Python 3
```
